### PR TITLE
Migrate nll_loss from TH to ATen (CPU)

### PR DESCRIPTION
**Pull request by andreaskoepf**
This is a port of the negative log likelihood TH loss implementation to ATen which is used by `torch.nn.functional.nll_loss()` for 2d inputs (N, C).

## Performance Impact

I measured no significant performance-difference of the port compared to the original implementation when using this [benchmark test script](https://gist.github.com/andreaskoepf/3c8e3698607773db2788dfd8885a9ed9).

### WITH PR applied:
```
CPU forward 1000 took 2.5290995836257935e-05
CPU forward 10000 took 5.757302278652787e-05
CPU forward 100000 took 0.0004873779835179448
CPU forward 1000000 took 0.0051894880016334355
CPU forward 10000000 took 0.026263039995683357
CPU forward TOTAL time 0.8068871730065439
CPU for- & backward 1000 took 0.00018794499919749796
CPU for- & backward 10000 took 0.0002642899926286191
CPU for- & backward 100000 took 0.0011828370043076575
CPU for- & backward 1000000 took 0.01250307000009343
CPU for- & backward 10000000 took 0.11453165800776333
CPU for- & backward TOTAL time 0.824805997981457
```

### Original TH version:
```
CPU forward 1000 took 2.1958985598757863e-05
CPU forward 10000 took 6.608400144614279e-05
CPU forward 100000 took 0.0004632119962479919
CPU forward 1000000 took 0.005477247992530465
CPU forward 10000000 took 0.02681165697867982
CPU forward TOTAL time 0.8073387439944781
CPU for- & backward 1000 took 0.00020634100656025112
CPU for- & backward 10000 took 0.00031720998231321573
CPU for- & backward 100000 took 0.0011843869870062917
CPU for- & backward 1000000 took 0.010876987013034523
CPU for- & backward 10000000 took 0.09893897600704804
CPU for- & backward TOTAL time 0.8271351839939598
```

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Review by ifedan — APPROVED** [review]

**Comment by andreaskoepf** [comment]
references #24744, #24743, #15870

**Comment by ifedan** [comment]
It seems like one of the unit test failed: AtenXlaTensorTest.TestNllLossBackward

**Review by ifedan — CHANGES_REQUESTED** [review]
It seems like one of the unit test failed: AtenXlaTensorTest.TestNllLossBackward

**Comment by andreaskoepf** [comment]
> It seems like one of the unit test failed: AtenXlaTensorTest.TestNllLossBackward

oops, I will look into it! thx.

**Comment by andreaskoepf** [comment]
@ifedan I am not able to reproduce the error. I checked the source multiple times ... it is not clear to how a 0-dim (scalar?) tensor should make it to an accessor-construction call. There are several other checkse that should catch this earlier and generate other error messages. Can I see somehow the source of the unit test or even execute it on my machine? Otherwise I have no idea how to further track the problem down? General question: Are these special tests for XLA Tensors that are never run for the other tensor types? It is quite strange that all other tests passed?

**Comment by andreaskoepf** [comment]
OK, found the failing test in the XLA repo: [test_aten_xla_tensor.cpp#L7513-L7539](https://github.com/pytorch/xla/blob/d9977c6e76c62f66ef1988ded50264c8c6a81571/test/cpp/test_aten_xla_tensor.cpp#L7513-L7539).

**Comment by andreaskoepf** [comment]
@ezyang I rebased and added new input checks. Now I wanted to see the xla-test output with my latest modifications but somehow only step `circleci: pytorch_xla_linux_xenial_py3_6_clang7_build` was executed and `pytorch_xla_linux_xenial_py3_6_clang7_test` is missing in the circle-ci checks-list after my latest push. Is there any way to enable the `pytorch_xla_linux_xenial_py3_6_clang7_test` step?

**Comment by andreaskoepf** [comment]
@ifedan @ezyang The XLA test passes now. It failed before because the backward pass expected a 1d tensor for the `total_weight` buffer tensor. This 1d tensor was always returned in a tuple together with the function result by the forward pass. It is unclear what causes the 1d->Scalar conversion. Some other code must convert the 1d tensor into a scalar tensor before calling the backward function. I think it is quite strange that this conversion happens and I have not yet found the code location where it is done. If you know where this happens please let me know - it is quite important for me to understand why and when automatic conversions of the forward-outputs happen, e.g. are single element 1d tensors always converted into scalar-tensors?

btw other tests fail now, which are unrelated to this pr, e.g. win build out of mem and 'ctc_loss leak problem' in pytorch_linux_xenial_cuda9_cudnn7_py3_NO_AVX_NO_AVX2_test .. if you could restart those tests it would be great (I have no permission to re-run the workflows) !

**Comment by ezyang** [comment]
It's hard to tell exactly what the problem is from your description, but the shapes of corresponding forwards and backwards should always match. I think the more likely problem is there is some sort of scalar v. 1-d tensor confusion in TH: historically, TH didn't understand scalar tensors, and there's a bunch of BC code for handling this case. But it's hard to say exactly for certain.

File: aten/src/ATen/native/LossNLL.cpp [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    int64_t ignore_index) {
+  const auto n_dims = input.dim();
+  const auto n_classes = input.size(-1);
+
+  scalar_t* total_weight_data = total_weight.data_ptr<scalar_t>();
+  *total_weight_data = 0;
+
+  auto weight_contiguous = weight.contiguous();
+  const scalar_t* weight_data =
+      weight.defined() ? weight_contiguous.data_ptr<scalar_t>() : nullptr;
```
Comment (ezyang) [inline aten/src/ATen/native/LossNLL.cpp]: If weight wasn't defined, wouldn't `weight.contiguous()` fail?

**Comment by ezyang** [comment]
@andreaskoepf So, I rereviewed your patch, and I see that you are now returning a scalar output. Is it possible that the XLA implementation was returning a 1-d tensor? The XLA is https://github.com/pytorch/xla/blob/master/torch_xla/csrc/nll_loss.cpp

File: aten/src/ATen/native/LossNLL.cpp (hunk 1 above)
Comment (andreaskoepf) [reply aten/src/ATen/native/LossNLL.cpp]: It worked during my tests but I just had a look at the impl and I think I have to retest in debug mode. By default the `is_contiguous_` member of a TensorImpl (also UndefinedTensor) should be true ... contiguous() itself does ```  if (self.is_contiguous(memory_format)) return self; } ``` .. but there is a
```
#ifdef DEBUG
  AT_ASSERT(compute_contiguous() == is_contiguous_);
#endif
``` 
in `TensorImpl::is_contiguous()` ... and I fear the compute_contiguous() might throw. I will check. Thanks for bringing that up.

**Comment by andreaskoepf** [comment]
> Is it possible that the XLA implementation was returning a 1-d tensor?

I do not think that this was a problem. The situation I noticed was that the buffer variable `total_weight` which was returned as 1D 1 element tensor from the forward function was passed as scalar tensor to the backward function. I have not yet found out why this happened and why it does not happen when backward is called from Python. The current code only requires that `total_weight.numel() == 1` and that the element is accessible via `*total_weight.data_ptr<scalar_t>()` ...

I hope in a future pass of refactoring we can also remove `total_weight` completely from the nll_loss impl since its value can be easily recomputed in the backward pass.

File: aten/src/ATen/native/LossNLL.cpp (hunk 1 above)
Comment (andreaskoepf) [reply aten/src/ATen/native/LossNLL.cpp]: I tested with DEBUG=1 build but apparently that does not define the DEBUG preprocessor symbol. From reviewing the source I believe that calling contiguous() on an undefined tensor breaks as soon as `DEBUG` is defined, therefore I added a helper function that checks if the tensor is defined before calling contiguous...

**Comment by ezyang** [comment]
But `total_weight` is a scalar, per https://github.com/pytorch/pytorch/pull/28270/files#diff-638f4441d15104ec8089e5c4041552caR163 (or maybe you're talking about an earlier version of the code?)

**Comment by andreaskoepf** [comment]
@ezyang I changed it to return a scalar in order to make the affected XLA unit test pass, in this [commit](https://github.com/pytorch/pytorch/pull/28270/commits/8d7390b163a2a5eb105368c44cf26b95f7600fe2). Before the code in the forward pass produced a 1D-tensor and and the code in the backward pass assumed that also a 1D tensor for total-weight would be passed. I have not yet invested the time to investigate why the original code did not work ONLY for the XLA tests .. e.g. all other 70+ checks before (which also should trigger the backward pass of the loss-function) passed. Maybe I will try if invoking from C++ is somehow special...
